src/status_im/subs/wallet/screens/input_amount.cljs

```diff
@@ -0,0 +1,360 @@
+ :<- [:send-input-amount-screen/token-decimals]
+ (fn [[{:keys [crypto-currency?]}
+       token-decimals]]
+   (if crypto-currency? token-decimals 2)))
+
+(rf/reg-sub :send-input-amount-screen/upper-limit
+ :<- [:send-input-amount-screen/state]
+ :<- [:send-input-amount-screen/conversion-rate]
+ :<- [:send-input-amount-screen/usd-conversion-rate]
+ :<- [:send-input-amount-screen/total-balance]
+ (fn [[{:keys [crypto-currency?]}
+       conversion-rate
+       usd-conversion-rate
+       total-balance]]
+   (if crypto-currency?
+     (utils/cut-crypto-decimals-to-fit-usd-cents
+      total-balance
+      usd-conversion-rate)
+     (utils/cut-fiat-balance-to-two-decimals
+      (money/crypto->fiat total-balance conversion-rate)))))
+
+(rf/reg-sub :send-input-amount-screen/upper-limit-prettified
+ :<- [:send-input-amount-screen/state]
+ :<- [:profile/currency-symbol]
+ :<- [:wallet/wallet-send-token]
+ :<- [:send-input-amount-screen/conversion-rate]
+ :<- [:send-input-amount-screen/upper-limit]
+ (fn [[{:keys [crypto-currency?]}
+       currency-symbol
+       {token-symbol :symbol}
+       conversion-rate
+       upper-limit]]
+   (if crypto-currency?
+     (utils/prettify-crypto-balance
+      (or (clj->js token-symbol) "")
+      (money/bignumber upper-limit)
+      conversion-rate)
+     (utils/prettify-balance currency-symbol
+                             (money/bignumber upper-limit)))))
+
+(rf/reg-sub :send-input-amount-screen/upper-limit-exceeded?
+ :<- [:send-input-amount-screen/state]
+ :<- [:send-input-amount-screen/upper-limit]
+ (fn [[{:keys [input-value]}
+       upper-limit]]
+   (controlled-input-logic/upper-limit-exceeded?
+    input-value
+    upper-limit)))
+
+(rf/reg-sub :send-input-amount-screen/value-out-of-limits?
+ :<- [:send-input-amount-screen/state]
+ :<- [:send-input-amount-screen/upper-limit]
+ (fn [[{:keys [input-value]}
+       upper-limit]]
+   (controlled-input-logic/value-out-of-limits? input-value upper-limit 0)))
+
+(rf/reg-sub :send-input-amount-screen/valid-input?
+ :<- [:send-input-amount-screen/state]
+ :<- [:send-input-amount-screen/value-out-of-limits?]
+ (fn [[{:keys [input-value]}
+       value-out-of-limits?]]
+   (not (or (controlled-input-logic/empty-value? input-value) value-out-of-limits?))))
+
+(rf/reg-sub :send-input-amount-screen/upper-limit-equals-input-value?
+ :<- [:send-input-amount-screen/state]
+ :<- [:send-input-amount-screen/upper-limit]
+ (fn [[{:keys [input-value]}
+       upper-limit]]
+   (money/equal-to
+    (money/bignumber input-value)
+    (money/bignumber upper-limit))))
+
+
+(defn- fiat->crypto
+  [value conversion-rate]
+  (-> value
+      (money/fiat->crypto conversion-rate)
+      (utils/cut-crypto-decimals-to-fit-usd-cents conversion-rate)))
+
+(defn- crypto->fiat
+  [value conversion-rate]
+  (-> value
+      (money/crypto->fiat conversion-rate)
+      (utils/cut-fiat-balance-to-two-decimals)))
+
+(rf/reg-sub :send-input-amount-screen/input-value-converted
```

Comment:
As a general good practice, I always thought the subs should return data for the view layer, but not formatted data because that's a UI concern. One reason is that subs that return formatted data are usually not reusable. A formatted number can't be easily reused to compute something.

Reply:
Not sure about this one @ilmotta 

Some node subs can't be reusable and that's fine, there can be some subs that are leafs that are screen-specific. The previous subs node (before formatting) can be made to be reused.

Otherwise, we would need to handle formatting logic in the view and it's not needed. Subs are cached and we aren't using reagent form-2 components, if we wanted to avoid the re-computation in the view we'd need to use a hook, which leads to something equal or more complex.

Reply:
@ulisesmac There's nothing inherently wrong with using subs to return formatted data, but in my experience it does come with some problems because the subs IMO should be more independent, let's say, more stable than views.

Translations, number formatting, date formatting, string concatenations for presentation, etc those things I personally prefer to leave to the view layer. I've seen subs returning props for view components (even anonymous functions), just one example where bringing responsibilities to the sub layer can lead to sub-optimal solutions. But that's why I said "As a general practice, I always thought" so it's more of a personal preference.

Though I have to say this practice of formatting only at the view layer is as old as software, or at least I remember from pre-2010 era where backend controllers would return only data to the view and the view would use its own utilities to format numbers, etc. This architecture made the controllers simpler, more stable, improved reusability, easier to test, similar idea I see with subs.

Performance of these final presentation transformations shouldn't be a problem, but sometimes they might be I agree. For example, translations are memoized. I (personally) don't like the approach of caching all transformations in the view layer as we started to do since we migrated to hooks everywhere. I'm gonna guess almost nobody is profiling (myself included). I've seen many PRs caching stuff that is literally instant in CLJS, like the result of a simple `case` macro. So you know, worrying about caching formatting logic seems unnecessary unless profiling proves otherwise.

Reply:
Thanks for your reply, I agree on your POV.